src/rest_framework_dso/embedding.py

```diff
@@ -0,0 +1,198 @@
+    auto_expand_all = fields_to_expand is True
+    if auto_expand_all:
+        fields_to_expand = allowed_names
+
+    for field_name in fields_to_expand:
+        if field_name not in allowed_names:
+            raise _expand_parse_error(allowed_names, field_name) from None
+
+        # Get the field
+        try:
+            field = getattr(parent_serializer, field_name)
+        except AttributeError:
+            raise RuntimeError(
+                f"{parent_serializer.__class__.__name__}.{field_name}"
+                f" does not refer to an embedded field."
+            ) from None
+
+        # Check access via scopes (NOTE: this is higher-level functionality)
+        scopes = fetch_scopes_for_model(field.related_model)
+        if auth_checker and not auth_checker(*scopes.table):
+            # Not allowed, silently drop for _expand=true request.
+            if auto_expand_all:
+                continue
+
+            # Explicitly mentioned, raise error.
+            raise PermissionDenied(f"Eager loading not allowed for field '{field_name}'")
+
+        embedded_fields[field_name] = field
+
+    return embedded_fields
+
+
+def _expand_parse_error(allowed_names, field_name):
+    """Generate the proper exception for the invalid expand name"""
+    available = ", ".join(sorted(allowed_names))
+    if not available:
+        return ParseError("Eager loading is not supported for this endpoint")
+    else:
+        return ParseError(
+            f"Eager loading is not supported for field '{field_name}', "
+            f"available options are: {available}"
+        )
+
+
+class IteratorObserver:
+    """Observe the objects that are being returned.
+
+    Unlike itertools.tee(), retrieved objects are directly processed by other functions.
+    As built-in feature, the number of returned objects is also counted.
+    """
+
+    def __init__(self, iterable, observers=None):
+        self.number_returned = 0
+        self._iterable = iter(iterable)
+        self._item_callbacks = list(observers) if observers else []
+        self._is_iterated = False
+        self._has_items = None
+
+    def add_observer(self, callback):
+        """Install an observer callback that is notified when items are iterated"""
+        self._item_callbacks.append(callback)
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        """Keep a count of the returned items, and allow to notify other generators"""
+        try:
+            value = next(self._iterable)
+        except StopIteration:
+            self._is_iterated = True
+            raise
+
+        self.number_returned += 1
+        self._has_items = True
+
+        # Notify observers
+        for notify_callback in self._item_callbacks:
+            notify_callback(value)
+
+        return value
+
+    def __len__(self):
+        if not self._is_iterated:
+            raise RuntimeError("Unable to read counting generator size yet")
```

Comment:
Beetje een terzijde, maar dit maakt bijv.

```python
it = IteratorObserver([1,2,3])
list(it)
```

onmogelijk, want list krijgt de RuntimeError. Zonder __len__ werkt dat wel. Dit is altijd het eerste wat ik doe als ik een iterator moet debuggen.

Reply:
ok, goede debugging 👍  Dit werkt idd niet, methode is weggehaald.